File: backend/app/routers/market.py
from fastapi import APIRouter

from app.services.commodity_service import CommodityService
from app.services.fred_service import FredService
from app.services.news_service import NewsService
from app.services.market_intelligence import MarketIntelligenceService

router = APIRouter()

commodity_service = CommodityService()
fred_service = FredService()
news_service = NewsService()
market_intel_service = MarketIntelligenceService()


@router.get("/commodities")
async def get_commodities():
    from app.services import commodity_store
    stored = commodity_store.load_all()
    current = commodity_service.get_current_prices(stored=stored)
    historical = commodity_service.get_historical_prices(stored=stored)
    return {"current": current, "historical": historical}


@router.post("/commodities/refresh")
async def refresh_commodities():
    """Force re-fetch all commodity data from yfinance/FRED and update Firestore."""
    result = commodity_service.refresh_data()
    return result


@router.get("/commodity-trends")
async def get_commodity_trends():
    """Return 3-month commodity price trend data."""
    return commodity_service.get_price_trends()


@router.get("/rates")
async def get_rates():
    return fred_service.get_current_rates()


@router.get("/news")
async def get_news():
    return news_service.get_news()


@router.get("/summary")
async def get_market_summary():
    return await market_intel_service.get_summary()

File: frontend/src/api/market.ts
import api from "./client";
import type { CommodityPrice, NewsItem, MarketSummary } from "../types/market";

export async function getCommodities(): Promise<{
  current: CommodityPrice[];
  historical: {
    monthly: Array<{ month: string; type: string; [commodity: string]: string | number }>;
    commodities: string[];
  };
}> {
  const { data } = await api.get("/api/market/commodities");
  return data;
}

export async function getRates(): Promise<Record<string, any>> {
  const { data } = await api.get("/api/market/rates");
  return data;
}

export async function getNews(): Promise<NewsItem[]> {
  const { data } = await api.get("/api/market/news");
  return data;
}

export async function getMarketSummary(): Promise<MarketSummary> {
  const { data } = await api.get("/api/market/summary");
  return data;
}

export async function refreshCommodities(): Promise<{ status: string; commodities: number; as_of: string }> {
  const { data } = await api.post("/api/market/commodities/refresh");
  return data;
}
